config/common.go

```diff
@@ -86,44 +98,65 @@ func (cs *commonStore) set(newCfg *model.Config, validate func(*model.Config) er
 //
 // This function assumes no lock has been acquired, as it acquires a write lock itself.
 func (cs *commonStore) load(f io.ReadCloser, needsSave bool, validate func(*model.Config) error, persist func(*model.Config) error) error {
+	// Split f into two so that we can have a configuration that does not have environment overrides
+	f2 := new(bytes.Buffer)
+	tee := io.TeeReader(f, f2)
+
 	allowEnvironmentOverrides := true
-	loadedCfg, environmentOverrides, err := unmarshalConfig(f, allowEnvironmentOverrides)
+	loadedCfg, environmentOverrides, err := unmarshalConfig(tee, allowEnvironmentOverrides)
 	if err != nil {
-		return errors.Wrapf(err, "failed to unmarshal config")
+		return errors.Wrapf(err, "failed to unmarshal config with env overrides")
+	}
+
+	// Keep track of the original values that the Environment settings overrode
+	loadedCfgNoEnv, _, err2 := unmarshalConfig(f2, false)
+	if err2 != nil {
+		return errors.Wrapf(err, "failed to unmarshal config without env overrides")
 	}
 
 	// SetDefaults generates various keys and salts if not previously configured. Determine if
 	// such a change will be made before invoking.
 	needsSave = needsSave || loadedCfg.SqlSettings.AtRestEncryptKey == nil || len(*loadedCfg.SqlSettings.AtRestEncryptKey) == 0
 	needsSave = needsSave || loadedCfg.FileSettings.PublicLinkSalt == nil || len(*loadedCfg.FileSettings.PublicLinkSalt) == 0
 	needsSave = needsSave || loadedCfg.EmailSettings.InviteSalt == nil || len(*loadedCfg.EmailSettings.InviteSalt) == 0
+	needsSave = needsSave || loadedCfgNoEnv.SqlSettings.AtRestEncryptKey == nil || len(*loadedCfgNoEnv.SqlSettings.AtRestEncryptKey) == 0
+	needsSave = needsSave || loadedCfgNoEnv.FileSettings.PublicLinkSalt == nil || len(*loadedCfgNoEnv.FileSettings.PublicLinkSalt) == 0
+	needsSave = needsSave || loadedCfgNoEnv.EmailSettings.InviteSalt == nil || len(*loadedCfgNoEnv.EmailSettings.InviteSalt) == 0
 
 	loadedCfg.SetDefaults()
+	loadedCfgNoEnv.SetDefaults()
 
 	if validate != nil {
 		if err = validate(loadedCfg); err != nil {
-			return errors.Wrap(err, "invalid config")
+			return errors.Wrap(err, "invalid config with env overrides")
+		}
+		if err = validate(loadedCfgNoEnv); err != nil {
+			return errors.Wrap(err, "invalid config without env overrides")
 		}
 	}
 
 	if changed := fixConfig(loadedCfg); changed {
 		needsSave = true
 	}
+	if changed := fixConfig(loadedCfgNoEnv); changed {
+		needsSave = true
+	}
 
 	cs.configLock.Lock()
 	var unlockOnce sync.Once
 	defer unlockOnce.Do(cs.configLock.Unlock)
 
+	oldCfg := cs.config
```

Comment:
Yes, that's a much nicer solution! 
I had to attach one function (removeEnvOverrides, new line 169) so that we can get at the commonStore's variables when we call it in the Set function, and the tests. The rest of the helpers (not attached to commonStore) are still below, is that the preferred organization?